aspnetcore/blazor/test.md

```diff
@@ -0,0 +1,137 @@
+---
+title: Test components in ASP.NET Core Blazor
+author: guardrex
+description: Learn how to test componments in Blazor apps.
+monikerRange: '>= aspnetcore-3.1'
+ms.author: riande
+ms.custom: mvc
+ms.date: 08/10/2020
+no-loc: [cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
+uid: blazor/test
+---
+# Test components in ASP.NET Core Blazor
+
+[Egil Hansen](https://egilhansen.com/)
+
+Testing is an important aspect of building stable and maintainable software.
+
+To test a Blazor component, the *Component Under Test* (CUT) is:
+
+* Rendered with relevant input for the test.
+* Depending on the type of test performed, possibly subject to interaction or modification. For example, event handlers can be triggered, such as an `onclick` event for a button.
+* Inspected for expected values.
+
+## Test approaches
+
+Two common approaches for testing Blazor components are end-to-end (E2E) testing and unit testing:
+
+* **Unit testing**: [Unit tests](/dotnet/core/testing/) are written with a unit testing library that provides:
+  * Component rendering.
+  * Inspection of component output and state.
+  * Triggering of event handlers and life cycle methods.
+  * Assertions that component behavior is correct.
+
+  [bUnit](https://github.com/egil/bUnit) is an example of a library that enables Razor component unit testing.
+
+* **E2E testing**: A test runner runs a Blazor app containing the CUT and automates a browser instance. The testing tool inspects and interacts with the CUT through the browser. [Selenium](https://github.com/SeleniumHQ/selenium) is an example of an E2E testing framework that can be used with Blazor apps.
+
+In unit testing, only the Blazor component (Razor/C#) is involved. External dependencies, such as services and JS interop, must be mocked. In E2E testing, the Blazor component and all of it's auxiliary infrastructure are part of the test, including CSS, JS, and the DOM and browser APIs.
+
+Test scope, how extensive the tests are, typically has an influence on the speed of the tests. Unit tests run on a subset of the app's subsystems and usually execute in milliseconds. E2E tests, which test a broad group of the app's subsystems, can take several seconds to complete.
```

Comment:
Yes ... I'll refactor that.

Reply:
If it helps...

The idea behind that sentence is to say something about why unit tests generally runs faster, because their scope is more focused, e.g. only touch the component, and doesn't have to deal with the browser, network, processes, etc.

Reply:
I just made it say ...

> *Test scope* describes how extensive the tests are. Test scope typically has an influence on the speed of the tests. Unit tests run on a subset of the app's subsystems and usually execute in milliseconds. E2E tests, which test a broad group of the app's subsystems, can take several seconds to complete. 

Look good?